metadetect/tests/test_lsst_skysub.py

```diff
@@ -0,0 +1,286 @@
+        star_catalog=stars,
+        coadd_dim=coadd_dim,
+        g1=0.02,
+        g2=0.00,
+        psf=psf,
+        se_dim=coadd_dim,
+    )
+    return sim_data
+
+
+def make_exp(dims):
+    import lsst.afw.image as afw_image
+    exp = afw_image.ExposureF(width=dims[1], height=dims[0])
+    exp.mask.addMaskPlane("BRIGHT")
+    return exp
+
+
+def show_mask(exp):
+    import lsst.afw.display as afw_display
+    display = afw_display.getDisplay(backend='ds9')
+    display.mtv(exp.mask)
+    input('hit a key')
+
+
+def show_image(exp):
+    import lsst.afw.display as afw_display
+    display = afw_display.getDisplay(backend='ds9')
+    display.mtv(exp)
+    display.scale('log', 'minmax')
+    input('hit a key')
+
+
+def check_skysub(meanvals, errvals, image_noise, true_sky=0):
+
+    meansky = meanvals.mean()
+    stdsky = meanvals.std()
+    errsky = stdsky/np.sqrt(meanvals.size)
+    errmean = errvals.mean()
+
+    tol = image_noise / 10
+
+    # we want the uncertainty on the mean sky to be small enough for a
+    # meaningful test
+    assert errsky < tol / 3
+
+    # our sims have no sky in them
+
+    print('image_noise:', image_noise)
+    print('tol:', tol)
+    print('true sky value is', true_sky)
+    print('mean of all trials: %g +/- %g' % (meansky, errsky))
+    print('sky error from trials:', stdsky)
+    print('mean predicted error:', errmean)
+
+    assert np.abs(meansky - true_sky) < tol
+
+
+def test_skysub_smoke():
+    seed = 5
+    rng = np.random.RandomState(seed)
+
+    dims = [1000, 1000]
+    noise = 1.0
+    skyval = 0.0
+
+    exp = make_exp(dims)
+    exp.image.array[:, :] = rng.normal(scale=noise, size=dims, loc=skyval)
+    exp.variance.array[:, :] = noise**2
+
+    lsst_skysub_mod.determine_and_subtract_sky(exp)
+
+    meta = exp.getMetadata()
+    assert 'BGMEAN' in meta
+    assert 'BGVAR' in meta
+
+
+def test_skysub_pure_noise():
+    """
+    check the mean sky over all trials is within a fraction of the noise level
+    """
+    seed = 5
+    rng = np.random.RandomState(seed)
+
+    dims = [1000, 1000]
+    noise = 1.0
+    skyval = 0.0
+
+    ntrial = 100
+    meanvals = np.zeros(ntrial)
+    errvals = np.zeros(ntrial)
+
+    for itrial in range(ntrial):
+        exp = make_exp(dims)
+        exp.image.array[:, :] = rng.normal(scale=noise, size=dims, loc=skyval)
+        exp.variance.array[:, :] = noise**2
+
+        lsst_skysub_mod.determine_and_subtract_sky(exp)
+
+        meta = exp.getMetadata()
+
+        meanvals[itrial] = meta['BGMEAN']
+        errvals[itrial] = np.sqrt(meta['BGVAR'])
+
+    check_skysub(meanvals, errvals, noise, true_sky=0)
+
+
+def test_skysub_sim_smoke():
+    seed = 812
+    rng = np.random.RandomState(seed)
+    sim = make_lsst_sim(rng, gal_type='fixed')
+
+    exp = sim['band_data']['i'][0]
+    lsst_skysub_mod.determine_and_subtract_sky(exp)
+
+    meta = exp.getMetadata()
+    assert 'BGMEAN' in meta
+    assert 'BGVAR' in meta
+
+
+def test_skysub_sim_fixed_gal():
```

Comment:
This test should be run with an additional non-zero sky val too.

Reply:
What would the test at the end be?  Because of iteration it will not equal the sky we put in

Reply:
Well, then the test has already failed. :(

Why are we iterating the sky subtraction?

Reply:
I suppose we could add a third iteration to see if that one is close to zero

Reply:
Why has it failed?

The iteration improves the accuracy of the sky subtraction and also greatly reduces the variation in the recovered sky

Reply:
The first time through, with objects in the image, it tends to overestimate the sky.  But after re-detecting and redoing sky determination it works much better

Reply:
And to answer your original question: the iteration was recommended by DM people

Reply:
> Because of iteration it will not equal the sky we put in

If we can't recover the sky we put in, that is an issue, right?

Reply:
On the second iteration it is subtracting a smaller amount than on the first, so the value of BGMEAN will be some number smaller in absolute value than the sky we put in.

Actually what we can do is keep a running total of the all the skys subtracted and check that is what we put in